File: frontend/src/utils/filamentMaterial.ts
export interface FilamentMaterialJson {
  family: string;
  subtype?: string | null;
  color_hex: string;
  profile_id?: string | null;
  setting_id?: string | null;
}

export interface FilamentMaterialInput {
  family?: string | null;
  subtype?: string | null;
  colorHex?: string | null;
  color_hex?: string | null;
  profileId?: string | null;
  profile_id?: string | null;
  settingId?: string | null;
  setting_id?: string | null;
  type?: string | null;
  color?: string | null;
  tray_type?: string | null;
  tray_sub_brands?: string | null;
  tray_color?: string | null;
  tray_info_idx?: string | null;
  rgba?: string | null;
  slicer_filament?: string | null;
  material?: FilamentMaterialJson | null;
}

const FILAMENT_TYPE_GROUPS = [
  ['PA-CF', 'PA12-CF', 'PAHT-CF'],
];

const EQUIV_MAP: Record<string, string> = {};
for (const group of FILAMENT_TYPE_GROUPS) {
  const canonical = group[0].toUpperCase();
  for (const entry of group) EQUIV_MAP[entry.toUpperCase()] = canonical;
}

const KNOWN_FAMILIES = [
  'PETG-HF', 'PETG HF', 'PETG-CF', 'PLA-CF', 'PAHT-CF', 'PA12-CF', 'PA6-CF', 'PA-CF',
  'NYLON', 'PETG', 'HIPS', 'PLA', 'ABS', 'ASA', 'TPU', 'PVA', 'PC', 'PA', 'PET',
].sort((a, b) => b.length - a.length);

const PROFILE_LABELS: Record<string, string> = {
  GFA00: 'Bambu PLA Basic',
  GFA01: 'Bambu PLA Matte',
  GFA02: 'Bambu PLA Metal',
  GFA05: 'Bambu PLA Silk',
  GFA06: 'Bambu PLA Silk+',
  GFA07: 'Bambu PLA Marble',
  GFA08: 'Bambu PLA Sparkle',
  GFA09: 'Bambu PLA Tough',
  GFA11: 'Bambu PLA Aero',
  GFA12: 'Bambu PLA Glow',
  GFA13: 'Bambu PLA Dynamic',
  GFA15: 'Bambu PLA Galaxy',
  GFA16: 'Bambu PLA Wood',
  GFA50: 'Bambu PLA-CF',
  GFG00: 'Bambu PETG Basic',
  GFG01: 'Bambu PETG Translucent',
  GFG02: 'Bambu PETG HF',
  GFG50: 'Bambu PETG-CF',
  GFG96: 'Generic PETG HF',
  GFG98: 'Generic PETG-CF',
  GFG99: 'Generic PETG',
  GFL95: 'Generic PLA High Speed',
  GFL96: 'Generic PLA Silk',
  GFL98: 'Generic PLA-CF',
  GFL99: 'Generic PLA',
  GFN03: 'Bambu PA-CF',
  GFN04: 'Bambu PAHT-CF',
  GFN05: 'Bambu PA6-CF',
  GFN98: 'Generic PA-CF',
  GFN99: 'Generic PA',
  GFU00: 'Bambu TPU 95A HF',
  GFU01: 'Bambu TPU 95A',
  GFU02: 'Bambu TPU for AMS',
  GFU98: 'Generic TPU for AMS',
  GFU99: 'Generic TPU',
};

function cleanToken(value?: string | null): string {
  return (value || '').replace('@', ' @ ').trim().split(/\s+/).filter(Boolean).join(' ');
}

function normalizeSubtype(value?: string | null): string | null {
  const subtype = cleanToken((value || '').split('@', 1)[0]);
  if (!subtype || ['GENERIC', 'BAMBU', 'BAMBU LAB'].includes(subtype.toUpperCase())) return null;
  if (subtype.startsWith('@')) return null;
  return subtype;
}

export function canonicalFilamentType(type?: string | null): string {
  const upper = (type || '').trim().toUpperCase();
  return EQUIV_MAP[upper] ?? upper;
}

export function normalizeColorHex(value?: string | null, fallback = '#808080FF'): string {
  let raw = (value || '').trim().replace(/^#/, '');
  if (raw.length !== 6 && raw.length !== 8) raw = fallback.replace(/^#/, '');
  if (raw.length === 6) raw += 'FF';
  if (!/^[0-9a-fA-F]{8}$/.test(raw)) {
    raw = fallback.replace(/^#/, '');
    if (raw.length === 6) raw += 'FF';
  }
  return `#${raw.slice(0, 8).toUpperCase()}`;
}

export function parseMaterialLabel(label?: string | null, familyHint?: string | null): { family: string; subtype: string | null } {
  const clean = cleanToken(label);
  const hint = cleanToken(familyHint).toUpperCase();
  let hintFamily = '';
  if (hint) {
    const normalizedHint = hint.replaceAll(' ', '-');
    hintFamily = KNOWN_FAMILIES.find((candidate) =>
      normalizedHint === candidate.toUpperCase().replaceAll(' ', '-')
    )?.toUpperCase() ?? '';
  }
  if (!clean && hint) return { family: hintFamily || hint, subtype: null };
  if (!clean) return { family: '', subtype: null };
  const upper = clean.toUpperCase();
  let family = '';
  let subtype: string | null = null;

  if (hintFamily && (upper === hintFamily || ` ${upper} `.includes(` ${hintFamily} `) || upper.startsWith(`${hintFamily}-`))) {
    family = hintFamily;
  } else {
    for (const candidate of KNOWN_FAMILIES) {
      const candidateUpper = candidate.toUpperCase();
      const padded = ` ${upper.replaceAll('-', ' - ')} `;
      const candidatePadded = ` ${candidateUpper.replaceAll('-', ' - ')} `;
      if (upper === candidateUpper || ` ${upper} `.includes(` ${candidateUpper} `) || padded.includes(candidatePadded)) {
        family = candidateUpper;
        break;
      }
    }
  }

  if (!family) {
    const [first, ...rest] = clean.split(' ');
    return { family: first.toUpperCase(), subtype: normalizeSubtype(rest.join(' ')) };
  }

  if (family === 'PETG-HF' || family === 'PETG HF') {
    family = 'PETG';
    subtype = 'HF';
  }

  const familyIdx = upper.indexOf(family);
  if (familyIdx >= 0) {
    subtype = normalizeSubtype(clean.slice(familyIdx + family.length).replace(/^[-\s]+/, '')) || subtype;
  } else if (family === 'PETG') {
    const idx = Math.max(upper.indexOf('PETG-HF'), upper.indexOf('PETG HF'));
    if (idx >= 0) subtype = normalizeSubtype(clean.slice(idx + 'PETG-HF'.length).replace(/^[-\s]+/, '')) || subtype;
  }

  return { family, subtype };
}

export function colorsAreSimilar(color1?: string | null, color2?: string | null, threshold = 40): boolean {
  const c1 = normalizeColorHex(color1);
  const c2 = normalizeColorHex(color2);
  if (c1.slice(7, 9) === '00' || c2.slice(7, 9) === '00') {
    return c1.slice(7, 9) === c2.slice(7, 9) && c1.slice(0, 7) === c2.slice(0, 7);
  }
  return [1, 3, 5].every((idx) =>
    Math.abs(parseInt(c1.slice(idx, idx + 2), 16) - parseInt(c2.slice(idx, idx + 2), 16)) <= threshold
  );
}

export function filamentMaterialIdentityKey(input: FilamentMaterial | FilamentMaterialInput | FilamentMaterialJson): string {
  const material = input instanceof FilamentMaterial ? input : new FilamentMaterial(input);
  return [
    material.compatibleFamilyKey,
    material.subtypeKey,
    material.colorHex,
  ].join('|');
}

export const FilamentMatchScore = {
  NoMatch: 0,
  Family: 100,
  MaterialSimilarColor: 200,
  MaterialColor: 300,
  Profile: 400,
} as const;

function profileMaterial(profileId?: string | null): { family: string; subtype: string | null } {
  const label = PROFILE_LABELS[(profileId || '').split('_')[0]];
  return label ? parseMaterialLabel(label) : { family: '', subtype: null };
}

export class FilamentMaterial {
  family: string;
  subtype: string | null;
  colorHex: string;
  profileId: string | null;
  settingId: string | null;

  constructor(input: FilamentMaterialInput) {
    const material = input.material;
    const profileId = cleanToken(input.profileId ?? input.profile_id ?? input.slicer_filament ?? input.tray_info_idx ?? material?.profile_id) || null;
    const base = parseMaterialLabel(
      [input.family ?? material?.family ?? input.type ?? input.tray_type ?? '', input.subtype ?? material?.subtype ?? ''].filter(Boolean).join(' '),
      input.family ?? input.type ?? input.tray_type ?? material?.family ?? undefined,
    );
    const prof = profileMaterial(profileId);
    this.family = (base.family || prof.family || '').toUpperCase();
    this.subtype = normalizeSubtype(input.subtype ?? material?.subtype) ?? base.subtype ?? prof.subtype;
    this.colorHex = normalizeColorHex(input.colorHex ?? input.color_hex ?? material?.color_hex ?? input.color ?? input.tray_color ?? input.rgba);
    this.profileId = profileId;
    this.settingId = cleanToken(input.settingId ?? input.setting_id ?? material?.setting_id) || null;
  }

  static fromAmsTray(tray: FilamentMaterialInput): FilamentMaterial {
    const parsed = parseMaterialLabel(tray.tray_sub_brands ?? tray.tray_type ?? tray.type, tray.tray_type ?? tray.type);
    return new FilamentMaterial({
      family: parsed.family,
      subtype: parsed.subtype,
      colorHex: tray.tray_color ?? tray.color,
      profileId: tray.tray_info_idx ?? tray.profile_id,
      settingId: tray.setting_id,
    });
  }

  static fromRequirement(req: FilamentMaterialInput): FilamentMaterial {
    return new FilamentMaterial(req);
  }

  static from3mfRequirement(req: FilamentMaterialInput): FilamentMaterial {
    return FilamentMaterial.fromRequirement(req);
  }

  static fromSpool(spool: FilamentMaterialInput): FilamentMaterial {
    return new FilamentMaterial({
      family: spool.family ?? spool.type,
      subtype: spool.subtype,
      colorHex: spool.rgba ?? spool.colorHex ?? spool.color_hex ?? spool.color,
      profileId: spool.slicer_filament ?? spool.profileId ?? spool.profile_id,
      settingId: spool.settingId ?? spool.setting_id,
    });
  }

  static fromQueueOverride(override: FilamentMaterialInput): FilamentMaterial {
    return new FilamentMaterial(override);
  }

  get familyKey(): string {
    return this.family.toUpperCase();
  }

  get compatibleFamilyKey(): string {
    return canonicalFilamentType(this.family);
  }

  get subtypeKey(): string {
    return (this.subtype || '').toUpperCase();
  }

  get rgbHex(): string {
    return this.colorHex.slice(0, 7);
  }

  get materialLabel(): string {
    return [this.family, this.subtype].filter(Boolean).join(' ');
  }

  get genericColorName(): string {
    if (this.colorHex.slice(7, 9) === '00') return 'Clear';
    const r = parseInt(this.colorHex.slice(1, 3), 16) / 255;
    const g = parseInt(this.colorHex.slice(3, 5), 16) / 255;
    const b = parseInt(this.colorHex.slice(5, 7), 16) / 255;
    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);
    const l = (max + min) / 2;
    let h = 0;
    let s = 0;
    if (max !== min) {
      const d = max - min;
      s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
      if (max === r) h = ((g - b) / d + (g < b ? 6 : 0)) / 6;
      else if (max === g) h = ((b - r) / d + 2) / 6;
      else h = ((r - g) / d + 4) / 6;
    }
    h *= 360;
    if (l < 0.15) return 'Black';
    if (l > 0.85) return 'White';
    if (s < 0.15) {
      if (l < 0.4) return 'Dark Gray';
      if (l > 0.6) return 'Light Gray';
      return 'Gray';
    }
    if (h >= 15 && h < 45 && l < 0.45) return 'Brown';
    if (h >= 45 && h < 70 && l < 0.40) return 'Brown';
    if (h < 15 || h >= 345) return 'Red';
    if (h < 45) return 'Orange';
    if (h < 70) return 'Yellow';
    if (h < 150) return 'Green';
    if (h < 200) return 'Cyan';
    if (h < 260) return 'Blue';
    if (h < 290) return 'Purple';
    return 'Pink';
  }

  get displayName(): string {
    return `${this.materialLabel || 'Unknown'} - ${this.genericColorName}`;
  }

  toQueueJson(): FilamentMaterialJson {
    return {
      family: this.family,
      subtype: this.subtype,
      color_hex: this.colorHex,
      profile_id: this.profileId,
      setting_id: this.settingId,
    };
  }

  toLegacyTypeColor(): { type: string; color: string; tray_info_idx: string } {
    return {
      type: this.family,
      color: this.rgbHex,
      tray_info_idx: this.profileId || '',
    };
  }

  toMqttFields(): {
    tray_type: string;
    tray_sub_brands: string;
    tray_color: string;
    tray_info_idx: string;
    setting_id: string;
  } {
    return {
      tray_type: this.family,
      tray_sub_brands: this.materialLabel || this.family,
      tray_color: this.colorHex.replace(/^#/, ''),
      tray_info_idx: this.profileId || '',
      setting_id: this.settingId || '',
    };
  }

  isFamilyMatch(other: FilamentMaterial): boolean {
    return !!this.compatibleFamilyKey && this.compatibleFamilyKey === other.compatibleFamilyKey;
  }

  isMaterialMatch(other: FilamentMaterial): boolean {
    if (!this.isFamilyMatch(other)) return false;
    return !this.subtypeKey || !other.subtypeKey || this.subtypeKey === other.subtypeKey;
  }

  isColorMatch(other: FilamentMaterial): boolean {
    return this.rgbHex.toUpperCase() === other.rgbHex.toUpperCase() && this.colorHex.slice(7, 9) === other.colorHex.slice(7, 9);
  }

  isSimilarColor(other: FilamentMaterial): boolean {
    return colorsAreSimilar(this.colorHex, other.colorHex);
  }

  isProfileMatch(other: FilamentMaterial): boolean {
    return !!this.profileId && !!other.profileId && this.profileId === other.profileId;
  }

  isExactMatch(other: FilamentMaterial): boolean {
    if (this.profileId && other.profileId) {
      return this.isProfileMatch(other);
    }
    return this.isMaterialMatch(other) && this.isColorMatch(other);
  }

  compatibilityScore(
    other: FilamentMaterial,
    policy: { allowFamilyFallback?: boolean } = {},
  ): (typeof FilamentMatchScore)[keyof typeof FilamentMatchScore] {
    if (this.isProfileMatch(other)) return FilamentMatchScore.Profile;
    if (this.isMaterialMatch(other) && this.isColorMatch(other)) return FilamentMatchScore.MaterialColor;
    if (this.isMaterialMatch(other) && this.isSimilarColor(other)) return FilamentMatchScore.MaterialSimilarColor;
    if (policy.allowFamilyFallback && this.isFamilyMatch(other)) return FilamentMatchScore.Family;
    return FilamentMatchScore.NoMatch;
  }
}

File: backend/app/services/filament_material.py
"""Canonical filament material value object.

This module keeps material identity separate from vendor colour names. Dispatch
compatibility is family-only; subtype, colour, and profile identifiers remain
available for display, strict colour policy, and printer protocol boundaries.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from backend.app.utils.filament_ids import filament_id_to_setting_id

_BRAND_PREFIXES = ("BAMBU LAB", "BAMBU", "GENERIC")


def canonical_filament_type(filament_type: str) -> str:
    """Return a normalized material family identifier.

    Families are intentionally never collapsed into equivalence groups. PAHT-CF,
    PA12-CF, and PA-CF are distinct printable materials.
    """
    return (filament_type or "").strip().upper()


def normalize_color_hex(value: str | None) -> str | None:
    """Normalize colour input to ``#RRGGBBAA``.

    Accepts ``#RRGGBB``, ``RRGGBB``, ``#RRGGBBAA``, or ``RRGGBBAA``. Missing
    alpha becomes ``FF``. Missing or invalid input remains unknown (``None``)
    rather than being converted into a real colour identity.
    """
    raw = (value or "").strip().lstrip("#")
    if len(raw) not in (6, 8):
        return None
    if len(raw) == 6:
        raw += "FF"
    try:
        int(raw, 16)
    except ValueError:
        return None
    return f"#{raw[:8].upper()}"


def _clean_token(value: str | None) -> str:
    return " ".join((value or "").replace("@", " @ ").split())


def _normalize_subtype(value: str | None) -> str | None:
    subtype = _clean_token((value or "").split("@", 1)[0])
    if not subtype or subtype.upper() in {"GENERIC", "BAMBU", "BAMBU LAB"}:
        return None
    if subtype.startswith("@"):
        return None
    return subtype


def _strip_brand_prefix(value: str) -> str:
    upper = value.upper()
    for prefix in _BRAND_PREFIXES:
        if upper == prefix:
            return ""
        if upper.startswith(f"{prefix} "):
            return value[len(prefix) :].strip()
    return value


def _canonical_family(value: str | None) -> str:
    return "-".join(_clean_token(value).upper().split())


def _split_family_and_subtype(value: str) -> tuple[str, str | None]:
    family, _, subtype = value.replace("-", " ").partition(" ")
    return _canonical_family(family), _normalize_subtype(subtype)


def parse_material_label(label: str | None, family_hint: str | None = None) -> tuple[str, str | None]:
    """Parse a slicer/printer material label into ``(family, subtype)``.

    An explicit source family (such as an AMS ``tray_type``) is authoritative.
    Labels only contribute optional subtype/display detail after a known brand
    prefix is removed. Without a source family, the first non-brand token is
    used as the family so newly introduced materials do not require code changes.
    """
    clean = _clean_token(label)
    display = _strip_brand_prefix(clean)
    hinted_family = ""
    hinted_subtype = None
    if family_hint:
        hint_display = _strip_brand_prefix(_clean_token(family_hint))
        hinted_family, hinted_subtype = _split_family_and_subtype(hint_display)

    if hinted_family:
        if not display:
            return hinted_family, hinted_subtype
        family_pattern = re.escape(hinted_family).replace(r"\-", r"[- ]+")
        match = re.search(rf"(?<!\w){family_pattern}(?!\w)", display, flags=re.IGNORECASE)
        subtype = _normalize_subtype(display[match.end() :].strip(" -")) if match else hinted_subtype
        return hinted_family, subtype or hinted_subtype
    if not display:
        return "", None
    return _split_family_and_subtype(display)


@dataclass(frozen=True)
class FilamentMaterial:
    family: str
    subtype: str | None
    color_hex: str | None
    profile_id: str | None = None
    setting_id: str | None = None

    def __post_init__(self) -> None:
        family = (self.family or "").strip().upper()
        subtype = _normalize_subtype(self.subtype)
        object.__setattr__(self, "family", family)
        object.__setattr__(self, "subtype", subtype)
        object.__setattr__(self, "color_hex", normalize_color_hex(self.color_hex))
        object.__setattr__(self, "profile_id", (self.profile_id or "").strip() or None)
        setting_id = (self.setting_id or "").strip() or None
        if not setting_id and self.profile_id:
            setting_id = filament_id_to_setting_id(self.profile_id)
        object.__setattr__(self, "setting_id", setting_id)

    @classmethod
    def from_parts(
        cls,
        *,
        family: str | None,
        subtype: str | None = None,
        color_hex: str | None = None,
        profile_id: str | None = None,
        setting_id: str | None = None,
    ) -> FilamentMaterial:
        parsed_family, parsed_subtype = parse_material_label(
            " ".join(part for part in [family or "", subtype or ""] if part).strip(),
            family if subtype is not None else None,
        )
        return cls(
            family=parsed_family or (family or ""),
            subtype=subtype if subtype is not None else parsed_subtype,
            color_hex=color_hex,
            profile_id=profile_id,
            setting_id=setting_id,
        )

    @classmethod
    def from_ams_tray(cls, tray: dict[str, Any]) -> FilamentMaterial:
        family_hint = tray.get("tray_type") or tray.get("type")
        family, subtype = parse_material_label(tray.get("tray_sub_brands") or family_hint, family_hint)
        return cls(
            family=family,
            subtype=subtype,
            color_hex=tray.get("tray_color") or tray.get("color"),
            profile_id=tray.get("tray_info_idx") or tray.get("profile_id"),
            setting_id=tray.get("setting_id"),
        )

    @classmethod
    def from_3mf_requirement(cls, req: dict[str, Any]) -> FilamentMaterial:
        material = req.get("material")
        if isinstance(material, dict):
            return cls.from_queue_material(material, fallback=req)
        family, subtype = parse_material_label(req.get("tray_sub_brands") or req.get("type"))
        return cls(
            family=family,
            subtype=subtype,
            color_hex=req.get("color") or req.get("color_hex"),
            profile_id=req.get("tray_info_idx") or req.get("profile_id"),
            setting_id=req.get("setting_id"),
        )

    @classmethod
    def from_spool(cls, spool: Any) -> FilamentMaterial:
        return cls.from_parts(
            family=getattr(spool, "material", None),
            subtype=getattr(spool, "subtype", None),
            color_hex=getattr(spool, "rgba", None),
            profile_id=getattr(spool, "slicer_filament", None),
        )

    @classmethod
    def from_queue_material(
        cls,
        material: dict[str, Any],
        *,
        fallback: dict[str, Any] | None = None,
    ) -> FilamentMaterial:
        fallback = fallback or {}
        return cls.from_parts(
            family=material.get("family") or fallback.get("type"),
            subtype=material.get("subtype"),
            color_hex=material.get("color_hex") or fallback.get("color"),
            profile_id=material.get("profile_id") or fallback.get("tray_info_idx"),
            setting_id=material.get("setting_id") or fallback.get("setting_id"),
        )

    @classmethod
    def from_queue_override(cls, override: dict[str, Any]) -> FilamentMaterial:
        material = override.get("material")
        if isinstance(material, dict):
            return cls.from_queue_material(material, fallback=override)
        return cls.from_parts(
            family=override.get("type") or override.get("family"),
            subtype=override.get("subtype"),
            color_hex=override.get("color") or override.get("color_hex"),
            profile_id=override.get("tray_info_idx") or override.get("profile_id"),
            setting_id=override.get("setting_id"),
        )

    @property
    def family_key(self) -> str:
        return self.family.upper()

    @property
    def compatible_family_key(self) -> str:
        return canonical_filament_type(self.family)

    @property
    def subtype_key(self) -> str:
        return (self.subtype or "").strip().upper()

    @property
    def rgb_hex(self) -> str | None:
        return self.color_hex[:7] if self.color_hex else None

    @property
    def material_label(self) -> str:
        return " ".join(part for part in [self.family, self.subtype] if part).strip()

    @property
    def generic_color_name(self) -> str:
        if not self.color_hex:
            return "Unknown colour"
        if self.color_hex[7:9] == "00":
            return "Clear"
        r = int(self.color_hex[1:3], 16) / 255
        g = int(self.color_hex[3:5], 16) / 255
        b = int(self.color_hex[5:7], 16) / 255
        max_v = max(r, g, b)
        min_v = min(r, g, b)
        lightness = (max_v + min_v) / 2
        saturation = 0.0
        hue = 0.0
        if max_v != min_v:
            delta = max_v - min_v
            saturation = delta / (2 - max_v - min_v) if lightness > 0.5 else delta / (max_v + min_v)
            if max_v == r:
                hue = ((g - b) / delta + (6 if g < b else 0)) / 6
            elif max_v == g:
                hue = ((b - r) / delta + 2) / 6
            else:
                hue = ((r - g) / delta + 4) / 6
        hue *= 360
        if lightness < 0.15:
            return "Black"
        if lightness > 0.85:
            return "White"
        if saturation < 0.15:
            if lightness < 0.4:
                return "Dark Gray"
            if lightness > 0.6:
                return "Light Gray"
            return "Gray"
        if 15 <= hue < 45 and lightness < 0.45:
            return "Brown"
        if 45 <= hue < 70 and lightness < 0.40:
            return "Brown"
        if hue < 15 or hue >= 345:
            return "Red"
        if hue < 45:
            return "Orange"
        if hue < 70:
            return "Yellow"
        if hue < 150:
            return "Green"
        if hue < 200:
            return "Cyan"
        if hue < 260:
            return "Blue"
        if hue < 290:
            return "Purple"
        return "Pink"

    @property
    def display_name(self) -> str:
        return f"{self.material_label or 'Unknown'} - {self.generic_color_name}"

    def to_queue_json(self) -> dict[str, str | None]:
        return {
            "family": self.family,
            "subtype": self.subtype,
            "color_hex": self.color_hex,
            "profile_id": self.profile_id,
            "setting_id": self.setting_id,
        }

    def to_api_json(self) -> dict[str, str | None]:
        """Serialize canonical identity together with backend-owned display text."""
        return {
            **self.to_queue_json(),
            "material_label": self.material_label,
            "display_name": self.display_name,
            "generic_color_name": self.generic_color_name,
        }

    def to_legacy_type_color(self) -> dict[str, str]:
        return {
            "type": self.family,
            "color": self.rgb_hex or "",
            "tray_info_idx": self.profile_id or "",
        }

    def to_mqtt_fields(self) -> dict[str, str]:
        return {
            "tray_type": self.family,
            "tray_sub_brands": self.material_label or self.family,
            "tray_color": (self.color_hex or "").lstrip("#"),
            "tray_info_idx": self.profile_id or "",
            "setting_id": self.setting_id or "",
        }

    def is_family_match(self, other: FilamentMaterial) -> bool:
        return bool(self.compatible_family_key and self.compatible_family_key == other.compatible_family_key)

    def is_material_match(self, other: FilamentMaterial) -> bool:
        """Return whether two materials have the same family and subtype identity."""
        return self.is_family_match(other) and self.subtype_key == other.subtype_key

    def is_dispatch_compatible(self, other: FilamentMaterial) -> bool:
        """Return whether a loaded material has the same material family.

        Dispatch deliberately ignores subtype and profile. They describe the
        selected printer/slicer settings, not whether PLA can be substituted
        for PLA. Colour policy is evaluated separately by the caller.
        """
        return self.is_family_match(other)

    def is_color_match(self, other: FilamentMaterial) -> bool:
        return bool(self.color_hex and other.color_hex and self.color_hex == other.color_hex)

    def is_profile_match(self, other: FilamentMaterial) -> bool:
        return bool(self.profile_id and other.profile_id and self.profile_id == other.profile_id)

    def is_exact_match(self, other: FilamentMaterial) -> bool:
        return self.is_material_match(other) and self.is_color_match(other)
